### PR TITLE
Added test cases, finished foreign key implementation

### PR DESCRIPTION
**Pull request by Hydrocharged**
Adds the entire foreign key test case suite from Dolt, in addition to cases originally missed that were discovered during the new implementation.

### CODE REVIEW

File: sql/plan/alter_foreign_key.go [hunk 1]
```diff
@@ -240,8 +256,25 @@ func resolveForeignKey(ctx *sql.Context, tbl sql.ForeignKeyTable, refTbl sql.For
 				Length: 0,
 			}
 		}
-		//TODO: generate a non-colliding index name
-		err := tbl.CreateIndexForForeignKey(ctx, fkDef.Name+"_idx", sql.IndexUsing_Default, sql.IndexConstraint_None, indexColumns)
+		indexMap := make(map[string]struct{})
```
Comment (andrew-wm-arthur) [inline sql/plan/alter_foreign_key.go:259]: add a helper method for this

File: sql/plan/alter_foreign_key.go [hunk 2]
```diff
@@ -269,10 +302,32 @@ func resolveForeignKey(ctx *sql.Context, tbl sql.ForeignKeyTable, refTbl sql.For
 			AppendTypes:    appendTypes,
 		},
 	}
+
+	// Check if the current foreign key name has already been used. Rather than checking the table first (which is the
+	// highest cost part of creating a foreign key), we'll check the name if it needs to be checked. If the foreign key
+	// was previously added, we don't need to check the name.
+	if shouldAdd {
```
Comment (andrew-wm-arthur) [inline sql/plan/alter_foreign_key.go:309]: helper method here

**Review by andrew-wm-arthur — APPROVED** [review]
LGTM, a few comments

**Comment by Hydrocharged** [comment]
Manually merged into base PR.

**Pull request closed without merging**